library/ssl_tls13_client.c

```diff
@@ -1611,16 +1611,22 @@ static int ssl_tls1_3_process_server_finished( mbedtls_ssl_context *ssl )
     if( ret != 0 )
         return( ret );
 
+    mbedtls_ssl_set_outbound_transform( ssl, ssl->handshake->transform_handshake );
```

Comment:
Looking at it, I think this should be done in ssl_tls13_finalize_server_hello at the same time where we change the inbound keys to the handshake one. That way if we have to send an alert message it will be done using the handshake keys. With early data in play, this story may be more complicated but for the time being that's all we need I think. Otherwise I would prefer to do that in #4965 with a specific commit.

Reply:
https://github.com/hannestschofenig/mbedtls/pull/379, I create a test for this suggestion in `prototype` . It breaks `ssl-opt.sh` . 
Please check what happens. 

And #4965 has included same change with this PR

Reply:
> That way if we have to send an alert message it will be done using the handshake key

Is that the right thing to do, though? If we have to send an alert after parsing the ServerHello, that means there was something wrong with it, so we don't have confidence that the server will actually be able to decrypt our alert?

Reply:
Is there anything fundamentally wrong with the current timing of setting the outgoing transform at the time of the ServerFinished?

Reply:
Otherwise, esp. if it conflicts with later PSK work, I would advice not to change it -- esp. because aligning with PSK isn't as far ahead as one might think, because we aim to keep `development` and `tls13-prototype` in sync.

Reply:
After looking at the change, I suspect it breaks the prototype because it's incompatible with the CCS Compatibility Mode: CCS Compatibility Mode demands sending an unencrypted (!) CCS after the ServerHello.

We've had a number of discussions around the placement of key changes and its ramifications, so unless we have a strong reason to change anything, I'd suggest we don't.

Reply:
For reference, the standard allows anything, but has a recommendation for handshake keys for post-SH alerts:

Reply:
@hanno-arm I only spotted an issue with the alert messages. I have done some testing and if after receiving the ServerHello we send an alert (could be while parsing the server Certificate message for example) not encrypted with the handshake key then the OpenSSL and GnuTLS servers reject the alert record.

Reply:
@ronald-cron-arm That's interesting -- I'm not sure if that's compliant on the side of OpenSSL and GnuTLS, I would have expected them to accept unencrypted alert messages at this point. 

What do you think of creating an issue considering to add this fast-forward to `handle_pending_alert()` at some point, but keeping the flow in the case of a successful handshake as-is?

Reply:
@hanno-arm Without considering any history of our code, to me, it is more natural to switch as done here the inbound and outbound key as soon as we are able to compute them and to have specific cases for early data and CCS.

Reply:
I disagree, it creates exceptions for the success-case, rather than exceptions for the failure case.

Reply:
Ok for compliant but not compatible with OpenSSL and GnuTLS.

Reply:
After going back to the spec I realize that changing back and forth the outbound key to send early data is just not possible. As the encrypted record all have the `application data` type the client and the server must be well synchronized on which key is currently in use. From the spec, it seems indeed fine to send not encrypted alerts thus I am ok with "creating an issue considering to add this fast-forward to handle_pending_alert() at some point" and with switching to the handshake key in `ssl_tls1_3_process_server_finished`. But this should be done in PR 4965, not here.